Improved slider

cchan987 opened the pull request:
viz.js 41-50 (Using zindex to move elements to front, allows click events on other elements) fixes #23

viz_sliders.js
30-33 (add a displayed value to the sliders) 
43-53 (sliders can only be used on mouseover) Fixes #12 
57-86 (sliders jump to middle/zero when middle clicked) Partial Fix #5 
92-131 (sliders can only be used on mouseover) Fixes #12

cchan987 commented:
I grouped these changes together as they were all connected, let me know if it would be better to split them into different pull requests/ branches

tcstewar commented:
Hmmm...  a few quick bug reports....

 - the z-index approach seems to mostly work, but I've seen a few situations where the slider graph isn't jumping to the front when I drag it around.  I haven't been able to quite reliably replicate the problem, though
 - Limiting slider movement on mouseover seems to have two undesirable consequences:
    - if you move quickly to the extreme limits of the slider, it ignores you
    - touch control doesn't work any more
 - Middle-click to reset-to-zero seems to only work about 10% of the time for me.

(for testing purposes, I was using Chrome under Windows)

cchan987 commented:
hmm, for the middle clicking, it should only works when clicking the darker grey slider bar.

cchan987 commented:
Added functionality for touch controls (needs testing on touchscreen which I don't have, @tcstewar may be able to help out XD)
and made the sliders more friendly for rapid dragging.

cchan987 commented:
zindex for putting elements on top updated, simpler, crisper, better.

cchan987 commented:
viz.css - changed background from white to transparent
viz.js - rewrote move element to top function using z-index

viz_slider.css - added slider guideline styling

viz_slider.js 
- Added the value to be displayed on the slider
- added zindex for the slider
-added , on slider drag end, if slider beyond limits, return drag-y value to boundary value
-added guidelines to the sliders
-modified resize handler to accommodate guideline

Point of concern:
-Guideline for sliders is a very long vertical line (999999) drawn in d3, when slider is resized, the svg holding the line is resized to show more/less of the line. I was struggling with redrawing the guideline without throwing the layout of the other elements off and this was a quick fix that presented itself.(thoughts?)

cchan987 commented:
guideline issue resolved

cchan987 commented:
Also note: the 2d.py example is in this branch, should it be removed?

tcstewar commented:
Looking pretty good...  :)

If you think the 2d.py example would be useful for future testing, then sure, leave it in.

The main thing I think still needs to get cleaned up here is that I'd like to keep with style of having spaces on either side of operators.  So ```i * width / N``` rather than ```i*width/N```.  Personally, I'm fine either way, but we should be consistent and all the existing code is with spaces.  :)

cchan987 commented:
Changes made as per discussion!

tcstewar commented on nengo_viz/static/viz_slider.js:
```diff
@@ -55,16 +97,56 @@ VIZ.Slider = function(args) {
+        //Good for positioning regardless of # of sliders
+        var guide_x = args.width / (2 * args.n_sliders) + 
+            (args.width / args.n_sliders) * i - this.guideline_width / 2;
+        VIZ.set_transform(guideline,guide_x,0);
```
Oops, forgot about this guideline:  spaces after commas

cchan987 commented:
whoops! fixed
